source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
Ok, on third (or fourth?) thought I think this is going to be complicated enough to merit its own PR (I'd like Matt taking a close look at that one).

I've changed it to a release assert, added a comment to fix in a follow-up, and changed the description here so it no longer closes the bug.

Given bind options are now back in the global proto I suggest we review this as is, do a follow-up to allow cluster overrides and a third follow-up for gentle bind fails.  WDYT?